### PR TITLE
Splunk events

### PR DESCRIPTION
**Pull request by robertbutacu**

### CODE REVIEW

File: app/assets/javascript/countdown.js [hunk 1]
```diff
@@ -24,9 +24,37 @@
         function myClock () {
           document.querySelector('.countdown').innerHTML = --c
           if (c === 0) {
-            clearInterval(config.myTimer)
-            var timeoutUrl = el.data('expiry-url')
-            window.location = timeoutUrl
+            var el = $('.countdown-wrapper')
+            var transitData = {
+              transitId: el.find('input[name="transit.transitId"]').val(),
```
Comment (udinobi) [inline app/assets/javascript/countdown.js]: Uhm... it looks really fragile this way to pass the Transit from the template to JS. However, speaking with Andrei, he is planning to refactor the entire Transit data assignment. So it is OK for the time being.

File: app/uk/gov/hmrc/ctc/controllers/TransitController.scala [hunk 2]
```diff
@@ -16,35 +16,22 @@
...
-import play.api.mvc.{Action, AnyContent}
+import play.api.libs.json.{Json, Reads}
+import play.api.mvc.{Action, AnyContent, Request, Result}
 import play.api.{Configuration, Environment}
 import uk.gov.hmrc.auth.core.Enrolment
 import uk.gov.hmrc.ctc.config.AppConfig
-import uk.gov.hmrc.ctc.services.TransitService
 import uk.gov.hmrc.ctc.connectors.StrideAuthConnector
+import uk.gov.hmrc.ctc.models.api.Transit
+import uk.gov.hmrc.ctc.models.submission.FormModels.UserData
```
Comment (udinobi) [inline app/uk/gov/hmrc/ctc/controllers/TransitController.scala]: The `submission` package looks to me not necessary. You can just have a package `form`.

File: app/uk/gov/hmrc/ctc/controllers/TransitController.scala [hunk 3]
```diff
@@ -16,35 +16,22 @@
...
+import uk.gov.hmrc.ctc.services.TransitService
+import uk.gov.hmrc.ctc.utils.Forms
```
Comment (udinobi) [inline app/uk/gov/hmrc/ctc/controllers/TransitController.scala]: ... and I'd move the `Forms` trait under `models.form`

File: app/uk/gov/hmrc/ctc/models/submission/FormModels.scala [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.ctc.models.submission
+
+import uk.gov.hmrc.ctc.models.api.Transit
+
+object FormModels {
```
Comment (udinobi) [inline app/uk/gov/hmrc/ctc/models/submission/FormModels.scala]: This seems to me not necessary.

File: app/uk/gov/hmrc/ctc/services/TransitService.scala [hunk 5]
```diff
@@ -20,20 +20,68 @@ import com.google.inject.Singleton
...
+    }
+
+    for {
+      nextTransit <- transitConnector.getNextTransitByPoEAndState(destinationPort, state)
+      _           <- auditResponse(nextTransit.map(_.transit))
+    } yield nextTransit
+  }
+
+  def complete(userData: UserData)(implicit headerCarrier: HeaderCarrier): Future[HttpResponse] ={
+    def createOfficeOfTransit(): Future[OfficeOfTransit] = Future {
```
Comment (udinobi) [inline app/uk/gov/hmrc/ctc/services/TransitService.scala]: Even though it looks like an elegant solution, it still seems to me that involve a Future for this simple computation is really overkill and unnecessary.

File: app/uk/gov/hmrc/ctc/views/transits.scala.html [hunk 6]
```diff
@@ -61,17 +61,26 @@
     <div class="ctc-text-right ctc-align-middle ctc-margin-top-2 countdown-wrapper" 
         data-module="countdown"
         data-countdown="@timeoutSeconds"
-        data-expiry-url="@TransitController.timeout(destinationPort, transitState)"
+        data-splunk-timeout-url="@uk.gov.hmrc.ctc.controllers.routes.TransitController.timeoutEvent(transitState)"
```
Comment (udinobi) [inline app/uk/gov/hmrc/ctc/views/transits.scala.html]: Please, import `uk.gov.hmrc.ctc.controllers.routes.TransitController`

File: app/uk/gov/hmrc/ctc/views/transits.scala.html [hunk 7]
```diff
@@ -61,17 +61,26 @@
...
+        data-timeout-url="@uk.gov.hmrc.ctc.controllers.routes.TransitController.timeout(destinationPort, transitState)"
```
Comment (udinobi) [inline app/uk/gov/hmrc/ctc/views/transits.scala.html]: Please, import `uk.gov.hmrc.ctc.controllers.routes.TransitController`

File: app/uk/gov/hmrc/ctc/views/transits.scala.html [hunk 8]
```diff
@@ -61,17 +61,26 @@
...
     >
-        <form class="countdown-form" method="POST" action="@TransitController.extendTimeout()" autocomplete="off" novalidate="novalidate">
-            <div class="form-group hidden">
-                <label for="csrfToken">CSRF Token</label>
+        <form class="countdown-form" method="POST" action="@uk.gov.hmrc.ctc.controllers.routes.TransitController.extendTimeout()" autocomplete="off" novalidate="novalidate">
```
Comment (udinobi) [inline app/uk/gov/hmrc/ctc/views/transits.scala.html]: Please, import `uk.gov.hmrc.ctc.controllers.routes.TransitController`

File: app/uk/gov/hmrc/ctc/controllers/TransitController.scala (hunk 2 above)
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/controllers/TransitController.scala]: Ah,yes - there were multiple a few forms before, but it has been refactored and only UserData was left

File: app/uk/gov/hmrc/ctc/services/TransitService.scala (hunk 5 above)
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: It is mostly for keeping the Monadic Flow in place - either way, the result will have to be lifted in a monadic context over Future, so either way or another, its the same result.

Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: Also, even though involving a Future might seem much, the Execution Context actually favours the current thread to execute it - so it's little difference to how the flow is impacted. Extracting that function outside the for comprehension ruins the established pattern

Comment (peppe-hmrc) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: I'm not sure you need to do `.map(_ => ())`. If I'm not wrong the `for` will not yield until `auditResponse` returns. 
The `for` is a sequence of map and the yield is a flatmap. Am I missing something?

Comment (peppe-hmrc) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: Also `Future` will create context swap and that future can't fail so is better to use `Future.successful()`

Comment (peppe-hmrc) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: And yes I also do not fully understand the use of the future to fit it in the `for`. 
`createOfficeOfTransit` should be a `val` in my opinion

Comment (peppe-hmrc) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: And also I would create an `apply(UserData)` for that :)

File: app/uk/gov/hmrc/ctc/services/TransitService.scala [hunk 9]
```diff
@@ -20,20 +20,68 @@ import com.google.inject.Singleton
...
+    } yield extendedTimeout
+  }
 
-  def complete(transitId: String, userData: UserData)(implicit headerCarrier: HeaderCarrier) =
-    transitConnector.complete(transitId, userData)
 
-  def extendTimeout(transitId: String)(implicit headerCarrier: HeaderCarrier): Future[TransitData] =
-    transitConnector.extendTimeout(transitId).map(response => TransitData(response.transit,response.timeoutSeconds))
+  def auditTimeout(transit: Transit)(implicit hc: HeaderCarrier): Future[Unit] = {
+    audit(ProcessingEvent.userTimedOut("some-pid", transit)).map(_ => ())
```
Comment (peppe-hmrc) [inline app/uk/gov/hmrc/ctc/services/TransitService.scala]: I'm not sure you need to `.map(_ => ())`

File: app/uk/gov/hmrc/ctc/services/TransitService.scala (hunk 5 above)
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: No, the whole for comprehension is a series of nested flatMaps and the yield is the map, hence, monads.

I didn't need the result of that call, so i thought of mapping it to Unit to express the side effect.

Future.successful sounds good - even though they most likely will get executed in the same thread.

File: app/uk/gov/hmrc/ctc/services/TransitService.scala (hunk 9 above)
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: wanted Future[Unit], audit() might return Disabled or some other results based on config files, we dont have any kind of logic for that now, so its useless

File: app/uk/gov/hmrc/ctc/services/TransitService.scala [hunk 10]
```diff
@@ -20,20 +20,68 @@ import com.google.inject.Singleton
...
 
 @Singleton
-class TransitService @Inject()(transitConnector: TransitConnector, appConfig: AppConfig)(implicit ec: ExecutionContext) {
+class TransitService @Inject()(transitConnector: TransitConnector,
+                               appConfig: AppConfig,
+                               val auditConnector: AuditConnector)
+                              (implicit val ec: ExecutionContext)  extends AuditEvents {
+
+  def getNextTransitByPoEAndState(destinationPort: String, state: String)(implicit headerCarrier: HeaderCarrier): Future[Option[TransitResponse]] = {
+    def auditResponse(transitMaybe: Option[Transit]): Future[Unit] = {
```
[the remark this answers is not in the document]
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: I dont find creating a SplunkEvent from those parameters a well defined pattern across the project to extract it into an apply function inside the ADT - its pretty specific, so id rather have it be part of the service

File: app/uk/gov/hmrc/ctc/services/TransitService.scala (hunk 5 above)
Comment (peppe-hmrc) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: My bad sorry sequence of flatmap and the last is a map. 
Future.successful is recommended form docs

File: app/uk/gov/hmrc/ctc/services/TransitService.scala (hunk 9 above)
Comment (peppe-hmrc) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: yes but still the .map is not needed. you can just ignore the return value

File: app/uk/gov/hmrc/ctc/services/TransitService.scala (hunk 5 above)
Comment (peppe-hmrc) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: The modification looks outdated so I say it again :) you don't need any Future here, that is a `val` and also as I suggested can be an `apply`

Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: Yes, changed to that - as far as the argument for val goes, id rather have it be part of the for comprehension - easier to read and understand imo

Comment (peppe-hmrc) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: so are you wrapping the `val` in a Future?

File: app/uk/gov/hmrc/ctc/services/TransitService.scala (hunk 9 above)
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: Yes, i agree - should anyone want to do anything with it, they should be able to get it - changed to reflect that

File: app/uk/gov/hmrc/ctc/services/TransitService.scala (hunk 5 above)
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: The function itself returns a Future - the way I see it, it's part of the service flow - what I wanted to achieve is a series of flatmaps representing the flow of the business logic. 

Simply used a Monad's ability to lift a regular value A into the monadic context so I can continue with the for comprehension

Comment (peppe-hmrc) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: Do you realise that is only unnecessary computation?
what is wrong with this implementation?
`    val createOfficeOfTransit =  OfficeOfTransit(userData)
    for {
      response        <- transitConnector.completeTransit(userData.transit.transitId, createOfficeOfTransit)
      _               <- audit(ProcessingEvent.userCompletedMRNProcessing("some-pid", userData.transit, createOfficeOfTransit))
    } yield response`

File: app/uk/gov/hmrc/ctc/controllers/TransitController.scala (hunk 3 above)
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/controllers/TransitController.scala]: this was done

File: app/uk/gov/hmrc/ctc/services/TransitService.scala (hunk 5 above)
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: The cost of wrapping and unwrapping in this particular case is close to nonexistent - should there be a multiple stacked monad transformer type and would this have been done dozens of time, I would see a case for the changes - even then the Hotspot Compiler might pick this pattern up and optimize it by creating a pool of objects for that stacked monad so performance might not be an issue.

But in this trivial case, it boils down to a matter of styling. 
There is nothing wrong with your implementation, same as there is nothing wrong with mine.

**Pull request merged**